metrics/multilabelmap.go

```diff
@@ -0,0 +1,245 @@
+		switch fv.Kind() {
+		case reflect.String:
+			fmt.Fprintf(&sb, "%s=%q", label, fv.String())
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			fmt.Fprintf(&sb, "%s=\"%d\"", label, fv.Int())
+		case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+			fmt.Fprintf(&sb, "%s=\"%d\"", label, fv.Uint())
+		case reflect.Bool:
+			fmt.Fprintf(&sb, "%s=\"%v\"", label, fv.Bool())
+		default:
+			panic(fmt.Sprintf("MultiLabelMap key field %q has unsupported type %v", ft.Name, fv.Type()))
+		}
+	}
+	sb.WriteString("}")
+	return sb.String()
+}
+
+// KeyValue represents a single entry in a [MultiLabelMap].
+type KeyValue[T comparable] struct {
+	Key   T
+	Value expvar.Var
+}
+
+func (v *MultiLabelMap[T]) String() string {
+	return `"MultiLabelMap"`
+}
+
+// WritePrometheus writes v to w in Prometheus exposition format.
+// The name argument is the metric name.
+func (v *MultiLabelMap[T]) WritePrometheus(w io.Writer, name string) {
+	if v.Type != "" {
+		io.WriteString(w, "# TYPE ")
+		io.WriteString(w, name)
+		io.WriteString(w, " ")
+		io.WriteString(w, v.Type)
+		io.WriteString(w, "\n")
+	}
+	if v.Help != "" {
+		io.WriteString(w, "# HELP ")
+		io.WriteString(w, name)
+		io.WriteString(w, " ")
+		io.WriteString(w, v.Help)
+		io.WriteString(w, "\n")
+	}
+	v.mu.RLock()
+	defer v.mu.RUnlock()
+
+	for _, kv := range v.sorted {
+		io.WriteString(w, name)
+		io.WriteString(w, kv.labels)
+		switch v := kv.val.(type) {
+		case *expvar.Int:
+			fmt.Fprintf(w, " %d\n", v.Value())
+		case *expvar.Float:
+			fmt.Fprintf(w, " %v\n", v.Value())
+		default:
+			fmt.Fprintf(w, " %s\n", kv.val)
+		}
+	}
+}
+
+// Init removes all keys from the map.
+func (v *MultiLabelMap[T]) Init() *MultiLabelMap[T] {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	v.sorted = nil
+	v.m.Range(func(k, _ any) bool {
+		v.m.Delete(k)
+		return true
+	})
+	return v
+}
+
+// addKeyLocked updates the sorted list of keys in v.keys.
+//
+// v.mu must be held.
+func (v *MultiLabelMap[T]) addKeyLocked(key T, val expvar.Var) {
+	ls := labelString(key)
+
+	ent := labelsAndValue[T]{key, ls, val}
+	// Using insertion sort to place key into the already-sorted v.keys.
+	i := sort.Search(len(v.sorted), func(i int) bool {
+		return v.sorted[i].labels >= ls
+	})
+	if i >= len(v.sorted) {
+		v.sorted = append(v.sorted, ent)
+	} else if v.sorted[i].key == key {
+		v.sorted[i].val = val
+	} else {
+		var zero labelsAndValue[T]
+		v.sorted = append(v.sorted, zero)
+		copy(v.sorted[i+1:], v.sorted[i:])
+		v.sorted[i] = ent
+	}
+}
+
+func (v *MultiLabelMap[T]) Get(key T) expvar.Var {
+	i, _ := v.m.Load(key)
+	av, _ := i.(expvar.Var)
+	return av
+}
+
+func newInt() expvar.Var   { return new(expvar.Int) }
+func newFloat() expvar.Var { return new(expvar.Float) }
+
+func (v *MultiLabelMap[T]) getOrFill(key T, fill func() expvar.Var) expvar.Var {
+	if v := v.Get(key); v != nil {
+		return v
+	}
+
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	if v := v.Get(key); v != nil {
```

Comment:
you call `Get` both with and without the mu, and it has no docs describing whether that's allowed

Reply:
because Get is exported and the mutex is not part of the public API. Same as https://pkg.go.dev/expvar#Map.Get